utils/oc.py

```diff
@@ -529,12 +535,15 @@ def init_oc_client(self, cluster_info):
                 jump_host = cluster_info.get('jumpHost')
             else:
                 jump_host = None
-            self.set_oc(
-                cluster,
-                OC(server_url, token, jump_host,
-                   settings=self.settings,
-                   init_projects=self.init_projects,
-                   init_api_resources=self.init_api_resources))
+            try:
+                oc_client = OC(server_url, token, jump_host,
+                               settings=self.settings,
+                               init_projects=self.init_projects,
+                               init_api_resources=self.init_api_resources)
+                self.set_oc(cluster, oc_client)
+            except StatusCodeError as e:
+                logging.error('The cluster is unreachable: error %s', str(e))
+                self.set_oc(cluster, '')
```

Comment:
We are getting out of options to signal something bad happened. We have `False` in the line 529 above, and we also the check for `None` in the `openshift_base` (which I don't see ever being set, but that's besided my point).
Using an empty string crosses a line that tells me we should actually have a class to signal a bad `OC` instance. That class should be used for all the cases (replacing `None`, `False` and `''`) and carry extra information that will be used in the openshift_base to select the log level, the log message and to decide whether or not to call `ri.register_error()`.

Reply:
I hope I can avoid writing a whole separate class for a simple catch like this one.  I mean in that case, I might as well set to False and just change the message would be easier TBH.

Reply:
A whole separate class is no big deal. In fact, by returning an empty string you're creating an instance of a string type class and setting its value to `''`. Not that far from what I proposed. Same happens to False, which is a subclass of the integer type with value 0.
Anyway, returning False would be fine, even though returning None would be more correct, but that would make things harder given the current code.

Reply:
i completely agree with @apahim.
@arilivigni we have talked about just this and you agreed. what has changed?

Reply:
I didn't remember us talking about a separate class just to catch the exception.  If I missed that please point me to where that was discussed?
I will implement a class then so it is more robust for what we are trying to achieve.  Thanks for the feedback.

Reply:
@apahim I added the oc_status class for exit statuses related to output choices of the oc command.  If I remove line 577 in oc it gives me an error `AttributeError: 'int' object has no attribute 'cleanup'`  I am sure maybe there is a better way around this so feedback welcome.

Reply:
That's exactly why I mentioned you needed a class. Now, even though you created a class, you're not really passing that new class around, you're just using it to select the right `int` and passing the `int` around. Meaning you can either have an `OC()` instance, which implements `cleanup()`, or an `int` instance, which doesn't.

The way out here is to check the type (with `isinstance()`) to decide whether to call `cleanup()` or not. Very non-pythonic. It would be fine to check types to select a code path, but not to decide which interfaces to use (see EAFP vs LBYL here: https://docs.python.org/3/glossary.html).

Anyway, I still think that you need to pass a custom class instance around (not just an `int`), that carries the error information and implements the minimum interfaces needed not to break the code. In this case, `__bool__` returning `False`, which will also be useful for the check in the `openshift_base.py`.

Reply:
I agree with you better to pass the class around.  I just made an update to do just that, but not sure if it is 100% what you were looking for.  It does allow to add values for the oc_status, but again open to feedback to make it better.

Reply:
You're still not passing a clustom class around... now you're back to passing `False`.

Sorry, I'd should have been more vebose here. What I mean is to have a class like this (created in the `utils/oc.py` module):

```python
class OCFail:
    def __init__(self, message):
        self.message = message

    def __bool__(self):
        return False
```

Then use it like this:

```python
        if automation_token is None:
            self.set_oc(cluster, OCFail(message='something something no automationToken'))
        else
            ...
            try:
                self.set_oc(cluster, OC(...))
            except StatusCodeError:
                self.set_oc(cluster, OCFail(message='something something unreachable'))
```

And check it (in the `openshift_base.py`) with:

```python
            if oc is None:
                msg = f"[{cluster}] cluster skipped."
                logging.debug(msg)
                continue
            if not oc:  # instead of "if oc is False"
                ri.register_error()
                logging.error(oc.message)
                continue
```

In that case, no changes will be required to the `OC_Map.cleanup()`, since it already checks with `if oc`. It will just work.

It can get more complex if you want, also replacing the `None` check by embedding more info into the `OCFail` instance, but that is what I meant from the beginning. 

Not saying this is the only way... ultimatelly, you tell me the approach you want to use and we go from there.

Reply:
No, you did state that I didn't understand it completely.  Thanks for the more verbose answer now I finally get it :100:

Reply:
Also cleanup still fails with this approach

Reply:
Ok @apahim thanks for the solution I believe this is the answer you were looking for.  I will also attempt the `None` case but just wanted to make sure this was more in line.

Reply:
If you instrument `oc` for the errors paths and pass an instance of `oc` to `OCFail`, will you need `OCFail` at all? In that case, all you need is to return `oc` and make `oc` return `False` on `bool(oc)` when something bad happens. As I said, you tell me the approach you want to use and we go from there.

Reply:
I want to handle all the error paths including `None` so if I remove `OCFail` can this be done directly in `init_oc_client`?  I thought that was the point of a separate class, unless that class is renamed and handles `False` and `None`.

Reply:
I think that what we are proposing here is the latter. To make `OCFail` more generic and cover all the cases where `OC()` won't be instantiated.
Meaning that instead of:
```python
            if oc is None:
                msg = f"[{cluster}] cluster skipped."
                logging.debug(msg)
                continue
            if not oc:
                ri.register_error()
                msg = f"[{cluster}] cluster has no automationToken."
                logging.error(msg)
                continue
```
you would have something like this pseudocode:
```python
            if not oc:
                if oc.am_i_a_bad_boy:
                    ri.register_error()            
                logging.log(level=oc.log_level, msg=oc.message)
                continue
```
But really, the current state is already good enough IMO. Feel free to consider the current code reviewable so we can get this merged and iterate later.

Reply:
Thanks so much @apahim I implemented this in my latest change.  Now I understand the need for extra info and how to use it.  Let me know your thought on this latest change.